src/main/java/uk/gov/hmcts/reform/api/deprecated/DeprecatedApiInterceptor.java

```diff
@@ -24,59 +25,44 @@
-            headers = getHeadersFromMethod(handlerMethod);
+        return !(handler instanceof HandlerMethod) || preHandle(response, (HandlerMethod) handler);
+    }
+
+    private boolean preHandle(HttpServletResponse response, HandlerMethod handler) {
+        for (GenericDeclaration source : getAnnotatedSources(handler)) {
+            Map<String, String> headers = getHeaders(source);
             if (headers != null) {
-                response.setHeader(WARNING, headers.get(WARNING));
-            } else {
-                log.trace("No @APIDeprecated headers found for request handler: {}", handler);
+                copyWarning(response, headers);
+                return true;
             }
-        } else {
-            response.setHeader(WARNING, headers.get(WARNING));
         }
+
+        log.trace("No @APIDeprecated headers found for request handler: {}", handler);
         return true;
     }
 
-    private Map<String, String> getHeadersFromClass(HandlerMethod handlerMethod) {
-        Class javaClass = handlerMethod.getBeanType();
-        Map<String, String> headers = responseHeadersByHandler.get(javaClass);
-        if (headers == null) {
-            APIDeprecated apiDeprecated = AnnotationUtils.findAnnotation(javaClass, APIDeprecated.class);
-            RequestMapping requestMapping = AnnotationUtils.findAnnotation(javaClass, RequestMapping.class);
-            if (apiDeprecated != null) {
-                headers = new HashMap<>();
-                headers.put(WARNING, getWarningMessage(apiDeprecated, requestMapping));
-                responseHeadersByHandler.put(javaClass, headers);
-                return headers;
-            }
-            return null;
-        } else {
-            return headers;
-        }
+    private static List<GenericDeclaration> getAnnotatedSources(HandlerMethod handler) {
+        return Arrays.asList(handler.getBeanType(), handler.getMethod());
     }
 
-    private Map<String, String> getHeadersFromMethod(HandlerMethod handlerMethod) {
-        Method javaMethod = handlerMethod.getMethod();
-        Map<String, String> headers = responseHeadersByHandler.get(javaMethod);
-        if (headers == null) {
-            APIDeprecated apiDeprecated = AnnotationUtils.findAnnotation(javaMethod, APIDeprecated.class);
-            RequestMapping requestMapping = AnnotationUtils.findAnnotation(javaMethod, RequestMapping.class);
-            if (apiDeprecated != null) {
-                headers = new HashMap<>();
-                headers.put(WARNING, getWarningMessage(apiDeprecated, requestMapping));
-                responseHeadersByHandler.put(javaMethod, headers);
-                return headers;
-            }
-            return null;
-        } else {
-            return headers;
+    private Map<String, String> getHeaders(GenericDeclaration source) {
+        return responseHeadersByHandler
+            .computeIfAbsent(source, o -> computeHeaders((GenericDeclaration) o));
```

Comment:
what's `o`?

Reply:
It's an argument to mapping function that was expressed as lambda.

Reply:
I mean... perhaps it would be good to give it a more meaningful name :)